src/fixed.c

```diff
@@ -9,42 +9,66 @@
 #define uint32_lo(i) ((i) & 0xffff)
 #define uint32_hi(i) ((i) >> 16)
 #define uint32_carry16 ((1) << 16)
+/*if (x > minx - epsilon && x < minx + epsilon) ...*/
+#define CHECK_INTERVAL(x, minx, epsilon) \
+    ((int32_t) ((x - (minx - epsilon)) | (minx + epsilon - x)) > 0)
 
 twin_fixed_t twin_fixed_sqrt(twin_fixed_t a)
 {
-    twin_fixed_t max = a, min = 0;
-
-    while (max > min) {
-        twin_fixed_t mid = (max + min) >> 1;
-        if (mid >= 181 * TWIN_FIXED_ONE) {
-            max = mid - 1;
-            continue;
-        }
-        twin_fixed_t sqr = twin_fixed_mul(mid, mid);
-        if (sqr == a)
-            return mid;
-        if (sqr < a)
-            min = mid + 1;
-        else
-            max = mid - 1;
+    if (a <= 0)
+        return 0;
+    
+    if (CHECK_INTERVAL(a, TWIN_FIXED_ONE, (1 << 8-1)))
+        return TWIN_FIXED_ONE;
+
+    // count leading zero
+    int offset = 0;
+    for (twin_fixed_t i = a; !(0x40000000 & i); i <<= 1) {
+        ++offset;
+    }
+    // shift left 'a' to expand more digit for sqrt precision
+    offset &=  ~1;
+    a <<= offset;
+    // calculate the digits need to shift back
+    offset >>= 1;
+    offset -= (16 >> 1);
+    //calculate sqrt
```

Comment:
/* Use digit-by-digit calculation to compute square root */
Use /**/ for comment.